pkg/serviceaccounts/oauthclient/oauthclientregistry.go

```diff
@@ -73,14 +235,125 @@ func (a *saOAuthClientAdapter) GetClient(ctx kapi.Context, name string) (*oautha
+func parseModelPrefixName(key string) (string, string, bool) {
+	for _, prefix := range modelPrefixes {
+		if strings.HasPrefix(key, prefix) {
+			return prefix, key[len(prefix):], true
+		}
+	}
+	return "", "", false
+}
+
+// extractRedirectURIs builds redirect URIs using the given models and namespace.
+// The returned redirect URIs may contain duplicates and invalid entries.
+func (a *saOAuthClientAdapter) extractRedirectURIs(modelsMap map[string]model, namespace string) redirectURIList {
+	var data redirectURIList
+	groupKindModelListMapper := map[unversioned.GroupKind]modelList{} // map of GroupKind to all models belonging to it
+	groupKindModelToURI := map[unversioned.GroupKind]namesToObjMapperFunc{
+		routeGroupKind: a.redirectURIsFromRoutes,
+		// TODO add support for ingresses by creating the appropriate GroupKind and namesToObjMapperFunc
+		// ingressGroupKind: a.redirectURIsFromIngresses,
+	}
+
+	for _, m := range modelsMap {
+		gk := m.getGroupKind()
+		if len(m.name) == 0 && gk == emptyGroupKind { // Is this a static redirect URI?
+			uri := redirectURI{} // No defaults wanted
+			uri.merge(&m)
+			data = append(data, uri)
+		} else if _, ok := groupKindModelToURI[gk]; ok { // a GroupKind is valid if we have a namesToObjMapperFunc to handle it
+			groupKindModelListMapper[gk] = append(groupKindModelListMapper[gk], m)
+		}
+	}
+
+	for gk, models := range groupKindModelListMapper {
+		if names := models.getNames(); names.Len() > 0 {
+			if objMapper := groupKindModelToURI[gk](namespace, names); len(objMapper) > 0 {
+				data = append(data, models.getRedirectURIs(objMapper)...)
+			}
+		}
+	}
+
+	return data
+}
+
+// redirectURIsFromRoutes is the namesToObjMapperFunc specific to Routes.
+// Returns a map of route name to redirect URIs that contain the default data as specified by the route's ingresses.
+func (a *saOAuthClientAdapter) redirectURIsFromRoutes(namespace string, osRouteNames sets.String) map[string]redirectURIList {
+	var routes []routeapi.Route
+	routeInterface := a.routeClient.Routes(namespace)
+	if osRouteNames.Len() > 1 {
+		if r, err := routeInterface.List(kapi.ListOptions{}); err == nil {
+			routes = r.Items
+		}
+	} else {
+		if r, err := routeInterface.Get(osRouteNames.List()[0]); err == nil {
+			routes = append(routes, *r)
+		}
+	}
+	routeMap := map[string]redirectURIList{}
+	for _, route := range routes {
+		if osRouteNames.Has(route.Name) {
+			routeMap[route.Name] = redirectURIsFromRoute(&route)
+		}
+	}
+	return routeMap
+}
+
+// redirectURIsFromRoute returns a list of redirect URIs that contain the default data as specified by the given route's ingresses.
+func redirectURIsFromRoute(route *routeapi.Route) redirectURIList {
+	var uris redirectURIList
+	uri := newDefaultRedirectURI()
+	uri.path = route.Spec.Path
+	if route.Spec.Port != nil {
+		uri.port = route.Spec.Port.TargetPort.String()
+	}
+	if route.Spec.TLS == nil {
+		uri.scheme = "http"
```

Comment:
splitting this logic between newDefaultRedirectURI() and here is hard to follow... an else that sets `"https"` would be better